openassessment/fileupload/api.py

```diff
@@ -29,3 +37,299 @@ def remove_file(key):
     Remove file from the storage
     """
     return backends.get_backend().remove_file(key)
+
+
+def get_student_file_key(student_item_dict, index=0):
+    """
+    Args:
+        student_item_dict: A dictionary containing keys ('student_id', 'course_id', 'item_id').
+        index (int, optional): The index of a file.
+    """
+    key_template = '{student_id}/{course_id}/{item_id}'
+    index = int(index)
+    if index > 0:
+        key_template += '/{index}'
+    return key_template.format(index=index, **student_item_dict)
+
+
+def _safe_load_json_list(field):
+    """
+    Tries to load JSON-ified string,
+    returns an empty list if we try to load some non-JSON-encoded string.
+    """
+    try:
+        return json.loads(field)
+    except ValueError:
+        return []
+
+
+class FileUpload(object):
+    """
+    A layer of abstraction over the various components of file
+    data stored as ORA XBlock user-scoped fields.
+    """
+    def __init__(self, name=None, description=None, size=None, index=None, descriptionless=False, **student_item_dict):
+        self.name = name
+        self.description = description
+        self.size = size
+        self.index = index
+        self.student_id = student_item_dict.get('student_id')
+        self.course_id = student_item_dict.get('course_id')
+        self.item_id = student_item_dict.get('item_id')
+        self.descriptionless = descriptionless
+
+    @property
+    def exists(self):
+        return (self.description is not None) or self.descriptionless
+
+    @property
+    def download_url(self):
+        if self.exists:
```

Comment:
What happens if self.exists == False ?

Reply:
Then `download_url` returns `None`, which is what we want - deleted files should have null as a download URL.